File: src/backend/schemas/operations.py
from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel


class FireStatus(str, Enum):
    ACTIVE = "ACTIVE"
    CONTAINED = "CONTAINED"
    FIRE_OUT = "FIRE_OUT"


class OperationCreate(BaseModel):
    fire_status: FireStatus
    start_time: datetime
    location: str
    size_hectares: Optional[float] = None
    notes: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    radius_meters: Optional[float] = None
    linked_report_ids: list[int] = []


class OperationUpdate(BaseModel):
    fire_status: Optional[FireStatus] = None
    start_time: Optional[datetime] = None
    location: Optional[str] = None
    size_hectares: Optional[float] = None
    notes: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    radius_meters: Optional[float] = None
    keep_overnight: Optional[bool] = None


class OperationResetPreview(BaseModel):
    archive_count: int
    carried_over_count: int


class OperationResetResponse(OperationResetPreview):
    reset_id: int


class OperationRestoreRequest(BaseModel):
    fire_status: FireStatus


class OperationLinkedReport(BaseModel):
    report_id: int
    status: str
    category: str
    sub_category: Optional[str] = None
    reported_at: Optional[datetime] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    trust_score: Optional[int] = None
    safety_status: Optional[str] = None
    reporting_context: Optional[str] = None
    linked_operation_id: Optional[int] = None
    linked_operation_label: Optional[str] = None
    distance_meters: Optional[float] = None


class OperationResponse(BaseModel):
    operation_id: int
    fire_status: FireStatus
    start_time: datetime
    location: str
    size_hectares: Optional[float]
    notes: Optional[str]
    created_by: Optional[uuid.UUID]
    created_at: datetime
    updated_at: datetime
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    radius_meters: Optional[float] = None
    is_archived: bool = False
    archived_at: Optional[datetime] = None
    archived_by: Optional[uuid.UUID] = None
    archive_reason: Optional[str] = None
    keep_overnight: bool = False
    carried_over_at: Optional[datetime] = None
    last_reset_at: Optional[datetime] = None
    linked_report_ids: list[int] = []
    linked_reports: list[OperationLinkedReport] = []

    class Config:
        from_attributes = True


class LinkableReportSearchResponse(OperationLinkedReport):
    link_disabled: bool = False
    disabled_reason: Optional[str] = None


class LinkReportRequest(BaseModel):
    report_id: int

File: src/frontend/src/lib/api/operations.ts
import { apiFetch } from './transport';

export type FireStatus = 'ACTIVE' | 'CONTAINED' | 'FIRE_OUT';

export interface Operation {
  operation_id: number;
  fire_status: FireStatus;
  start_time: string;
  location: string;
  size_hectares: number | null;
  notes: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
  latitude: number | null;
  longitude: number | null;
  radius_meters: number | null;
  linked_report_ids: number[];
  linked_reports: LinkedReportDetail[];
}

export interface OperationCreate {
  fire_status: FireStatus;
  start_time: string;
  location: string;
  size_hectares?: number;
  notes?: string;
  latitude?: number;
  longitude?: number;
  radius_meters?: number;
  linked_report_ids?: number[];
}

export interface OperationUpdate {
  fire_status?: FireStatus;
  start_time?: string;
  location?: string;
  size_hectares?: number;
  notes?: string;
  latitude?: number;
  longitude?: number;
  radius_meters?: number;
}

export async function fetchOperations(status?: FireStatus[]): Promise<Operation[]> {
  const params = new URLSearchParams();
  status?.forEach((s) => params.append('status', s));
  const qs = params.toString();
  return apiFetch<Operation[]>(`/operations${qs ? `?${qs}` : ''}`);
}

export async function createOperation(data: OperationCreate): Promise<Operation> {
  return apiFetch<Operation>('/operations', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

export async function updateOperation(id: number, data: OperationUpdate): Promise<Operation> {
  return apiFetch<Operation>(`/operations/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  });
}

export async function deleteOperation(id: number): Promise<void> {
  return apiFetch<void>(`/operations/${id}`, { method: 'DELETE' });
}

export async function linkReport(operationId: number, reportId: number): Promise<Operation> {
  return apiFetch<Operation>(`/operations/${operationId}/link`, {
    method: 'POST',
    body: JSON.stringify({ report_id: reportId }),
  });
}

export async function unlinkReport(operationId: number, reportId: number): Promise<Operation> {
  return apiFetch<Operation>(`/operations/${operationId}/link/${reportId}`, {
    method: 'DELETE',
  });
}

export interface LinkedReportDetail {
  report_id: number;
  status: 'PENDING' | 'UNDER_REVIEW' | 'LINKED' | 'ACTIONED' | string;
  category: string;
  sub_category: string | null;
  reported_at: string | null;
  latitude: number | null;
  longitude: number | null;
  trust_score: number | null;
  safety_status: string | null;
  reporting_context: string | null;
  linked_operation_id: number | null;
  linked_operation_label: string | null;
  distance_meters: number | null;
}

export interface LinkableReportDetail extends LinkedReportDetail {
  link_disabled: boolean;
  disabled_reason: string | null;
}

export interface LinkableReportSearchParams {
  operation_id?: number;
  q?: string;
  status?: string[];
  category?: string;
  start?: string;
  end?: string;
  latitude?: number;
  longitude?: number;
}

export async function fetchLinkableReports(
  params: LinkableReportSearchParams = {},
): Promise<LinkableReportDetail[]> {
  const search = new URLSearchParams();
  if (params.operation_id != null) search.set('operation_id', String(params.operation_id));
  if (params.q) search.set('q', params.q);
  params.status?.forEach((s) => search.append('status', s));
  if (params.category) search.set('category', params.category);
  if (params.start) search.set('start', params.start);
  if (params.end) search.set('end', params.end);
  if (params.latitude != null) search.set('latitude', String(params.latitude));
  if (params.longitude != null) search.set('longitude', String(params.longitude));
  const qs = search.toString();
  return apiFetch<LinkableReportDetail[]>(`/operations/linkable-reports${qs ? `?${qs}` : ''}`);
}
